File: ArDemo/ARKController/SceneKit/AnchorNode.m
#import "AnchorNode.h"
#import "ARKHelper.h"

typedef NS_ENUM(NSUInteger, AnchorFigure)
{
    AnchorBox,
    AnchorSphere,
    AnchorPyramid,
    AnchorCylinder,
    AnchorCone,
    AnchorTube,
    AnchorCapsule,
    AnchorTorus,
    AnchorFugureCount
};

@interface AnchorNode ()
@property AnchorFigure figure;
@end

@implementation AnchorNode

- (void)dealloc
{
    DDLogDebug(@"- anchor");
}

- (instancetype)initWithAnchor:(ARAnchor *)anchor
{
    self = [super init];
    
    DDLogDebug(@"+ anchor");
    
    if (self)
    {
        _figure = AnchorBox;//arc4random() % AnchorPyramid;
        
        switch (_figure)
        {
            case AnchorBox:
                [self setupBox];
                break;
            case AnchorPyramid:
                [self setupPiramide];
                break;
            case AnchorSphere:
                [self setupSphere];
                break;
            case AnchorCylinder:
                [self setupCylinder];
                break;
            case AnchorCone:
                [self setupCone];
                break;
            case AnchorTube:
                [self setupTube];
                break;
            case AnchorCapsule:
                [self setupCapsule];
                break;
            case AnchorTorus:
                [self setupTorus];
                break;
            default:
                break;
        }
        
        [self setOpacity:0];
    }
    
    return self;
}

- (CGFloat)size
{
    switch (_figure)
    {
        case AnchorBox:
            return [(SCNBox *)[self geometry] width];
        case AnchorPyramid:
            return [(SCNPyramid *)[self geometry] width];
        case AnchorSphere:
            return [(SCNSphere *)[self geometry] radius];
        case AnchorCylinder:
            return [(SCNCylinder *)[self geometry] height];
        case AnchorCone:
            return [(SCNCone *)[self geometry] height];
        case AnchorTube:
            return [(SCNTube *)[self geometry] height];
        case AnchorCapsule:
            return [(SCNCapsule *)[self geometry] height];
        case AnchorTorus:
            return [(SCNTorus *)[self geometry] pipeRadius];
        default:
            break;
    }
    
    return 0;
}

- (void)setupTorus
{
    SCNGeometry *geometry = [SCNTorus torusWithRingRadius:BOX_SIZE pipeRadius:BOX_SIZE / 3];
    
    UIColor *materialColor;
    NSUInteger col = rand() % 7;
    switch (col)
    {
        case 0:
            materialColor = [UIColor greenColor];
            break;
        case 1:
            materialColor = [UIColor redColor];
            break;
        case 2:
            materialColor = [UIColor blueColor];
            break;
        case 3:
            materialColor = [UIColor greenColor];
            break;
        case 4:
            materialColor = [UIColor yellowColor];
            break;
        case 5:
            materialColor = [UIColor purpleColor];
            break;
        case 6:
            materialColor = [UIColor magentaColor];
            
            break;
        default:
            break;
    }

    geometry.materials = @[[self anchorMaterialWithColor:materialColor]];
    
    [self setGeometry:geometry];
}

- (void)setupCapsule
{
    SCNGeometry *geometry = [SCNCapsule capsuleWithCapRadius:BOX_SIZE height:BOX_SIZE];
    
    UIColor *materialColor;
    
    NSUInteger col = rand() % 7;
    switch (col)
    {
        case 0:
            materialColor = [UIColor greenColor];
            break;
        case 1:
            materialColor = [UIColor redColor];
            break;
        case 2:
            materialColor = [UIColor blueColor];
            break;
        case 3:
            materialColor = [UIColor greenColor];
            break;
        case 4:
            materialColor = [UIColor yellowColor];
            break;
        case 5:
            materialColor = [UIColor purpleColor];
            break;
        case 6:
            materialColor = [UIColor magentaColor];
            break;
        default:
            break;
    }
    
    geometry.materials = @[[self anchorMaterialWithColor:materialColor]];
    
    [self setGeometry:geometry];
}

- (void)setupTube
{
    SCNGeometry *geometry = [SCNTube tubeWithInnerRadius:(BOX_SIZE - BOX_SIZE / 5) outerRadius:BOX_SIZE height:BOX_SIZE];
    
    geometry.materials =  @[
                            [self anchorMaterialWithColor:[UIColor yellowColor]],
                            [self anchorMaterialWithColor:[UIColor purpleColor]],
                            [self anchorMaterialWithColor:[UIColor magentaColor]]];
    
    [self setGeometry:geometry];
}

- (void)setupCone
{
    SCNGeometry *geometry = [SCNCone coneWithTopRadius:BOX_SIZE/2 bottomRadius:BOX_SIZE height:BOX_SIZE];
    
    geometry.materials =  @[
                            [self anchorMaterialWithColor:[UIColor redColor]],
                            [self anchorMaterialWithColor:[UIColor greenColor]],
                            [self anchorMaterialWithColor:[UIColor blueColor]]];
    
    [self setGeometry:geometry];
}

- (void)setupCylinder
{
    SCNGeometry *geometry = [SCNCylinder cylinderWithRadius:BOX_SIZE height:BOX_SIZE];
    
    geometry.materials =  @[
                            [self anchorMaterialWithColor:[UIColor orangeColor]],
                            [self anchorMaterialWithColor:[UIColor grayColor]],
                            [self anchorMaterialWithColor:[UIColor blueColor]]];
    
    [self setGeometry:geometry];
}

- (void)setupSphere
{
    SCNGeometry *geometry = [SCNSphere sphereWithRadius:BOX_SIZE];
    
    UIColor *materialColor;
    
    NSUInteger col = rand() % 7;
    switch (col)
    {
        case 0:
            materialColor = [UIColor greenColor];
            break;
        case 1:
            materialColor = [UIColor redColor];
            break;
        case 2:
            materialColor = [UIColor blueColor];
            break;
        case 3:
            materialColor = [UIColor greenColor];
            break;
        case 4:
            materialColor = [UIColor yellowColor];
            break;
        case 5:
            materialColor = [UIColor purpleColor];
            break;
        case 6:
            materialColor = [UIColor magentaColor];
            break;
        default:
            break;
    }
    
    geometry.materials = @[[self anchorMaterialWithColor:materialColor]];
    
    [self setGeometry:geometry];
}

- (void)setupPiramide
{
    SCNGeometry *geometry = [SCNPyramid pyramidWithWidth:BOX_SIZE height:BOX_SIZE length:BOX_SIZE];
    
    geometry.materials =  @[
                            [self anchorMaterialWithColor:[UIColor redColor]],
                            [self anchorMaterialWithColor:[UIColor greenColor]],
                            [self anchorMaterialWithColor:[UIColor blueColor]],
                            [self anchorMaterialWithColor:[UIColor yellowColor]],
                            [self anchorMaterialWithColor:[UIColor purpleColor]]];
    
    [self setGeometry:geometry];
}

- (void)setupBox
{
    SCNGeometry *geometry = [SCNBox boxWithWidth:BOX_SIZE height:BOX_SIZE length:BOX_SIZE chamferRadius:(BOX_SIZE / 20)];
    
    geometry.materials =  @[
                            [self anchorMaterialWithColor:[UIColor redColor]],
                            [self anchorMaterialWithColor:[UIColor greenColor]],
                            [self anchorMaterialWithColor:[UIColor blueColor]],
                            [self anchorMaterialWithColor:[UIColor yellowColor]],
                            [self anchorMaterialWithColor:[UIColor purpleColor]],
                            [self anchorMaterialWithColor:[UIColor magentaColor]]];
    
    [self setGeometry:geometry];
}

- (SCNMaterial *)anchorMaterialWithColor:(UIColor *)color
{
    SCNMaterial *material = [SCNMaterial material];
    [material setFillMode:SCNFillModeFill];
    [[material diffuse] setContents:color];
    [[material specular] setContents:[UIColor whiteColor]];
    [material setLocksAmbientWithDiffuse:YES];
    
    return material;
}

@end

File: XRViewer/ARKController/SceneKit/AnchorNode.swift
import ARKit
import SceneKit
import XCGLogger

enum AnchorFigure : Int {
    case anchorBox
    case anchorSphere
    case anchorPyramid
    case anchorCylinder
    case anchorCone
    case anchorTube
    case anchorCapsule
    case anchorTorus
    case anchorFugureCount
}

class AnchorNode: SCNNode {
    
    private var figure: AnchorFigure?
    let boxSize = Constant.boxSize()
    
    init(anchor: ARAnchor) {
        super.init()

        appDelegate().logger.debug("+ anchor")
        
        figure = .anchorBox //arc4random() % AnchorPyramid;

        switch figure {
            case .anchorBox?:
                setupBox()
            case .anchorPyramid?:
                setupPyramid()
            case .anchorSphere?:
                setupSphere()
            case .anchorCylinder?:
                setupCylinder()
            case .anchorCone?:
                setupCone()
            case .anchorTube?:
                setupTube()
            case .anchorCapsule?:
                setupCapsule()
            case .anchorTorus?:
                setupTorus()
            default:
                break
        }

        opacity = 0
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    deinit {
        appDelegate().logger.debug("- anchor")
    }

    func size() -> CGFloat {
        switch figure {
            case .anchorBox?:
                return (geometry as? SCNBox)?.width ?? 0.0
            case .anchorPyramid?:
                return (geometry as? SCNPyramid)?.width ?? 0.0
            case .anchorSphere?:
                return (geometry as? SCNSphere)?.radius ?? 0.0
            case .anchorCylinder?:
                return (geometry as? SCNCylinder)?.height ?? 0.0
            case .anchorCone?:
                return (geometry as? SCNCone)?.height ?? 0.0
            case .anchorTube?:
                return (geometry as? SCNTube)?.height ?? 0.0
            case .anchorCapsule?:
                return (geometry as? SCNCapsule)?.height ?? 0.0
            case .anchorTorus?:
                return (geometry as? SCNTorus)?.pipeRadius ?? 0.0
            default:
                break
        }
        return 0
    }

    func setupTorus() {
        let geometry: SCNGeometry? = SCNTorus(ringRadius: boxSize, pipeRadius: boxSize / 3)

        var materialColor: UIColor?
        let col = Int(arc4random()) % 7
        switch col {
            case 0:
                materialColor = UIColor.green
            case 1:
                materialColor = UIColor.red
            case 2:
                materialColor = UIColor.blue
            case 3:
                materialColor = UIColor.green
            case 4:
                materialColor = UIColor.yellow
            case 5:
                materialColor = UIColor.purple
            case 6:
                materialColor = UIColor.magenta
            default:
                break
        }

        guard let anchorMaterial = anchorMaterial(with: materialColor) else { return }
        geometry?.materials = [anchorMaterial]

        self.geometry = geometry
    }

    func setupCapsule() {
        let geometry: SCNGeometry? = SCNCapsule(capRadius: boxSize, height: boxSize)

        var materialColor: UIColor?

        let col = Int(arc4random()) % 7
        switch col {
            case 0:
                materialColor = UIColor.green
            case 1:
                materialColor = UIColor.red
            case 2:
                materialColor = UIColor.blue
            case 3:
                materialColor = UIColor.green
            case 4:
                materialColor = UIColor.yellow
            case 5:
                materialColor = UIColor.purple
            case 6:
                materialColor = UIColor.magenta
            default:
                break
        }

        guard let anchorMaterial = anchorMaterial(with: materialColor) else { return }
        geometry?.materials = [anchorMaterial]

        self.geometry = geometry
    }

    func setupTube() {
        let geometry: SCNGeometry? = SCNTube(innerRadius: boxSize - boxSize / 5, outerRadius: boxSize, height: boxSize)

        guard let anchorMaterialYellow = anchorMaterial(with: .yellow) else { return }
        guard let anchorMaterialPurple = anchorMaterial(with: .purple) else { return }
        guard let anchorMaterialMagenta = anchorMaterial(with: .magenta) else { return }
        geometry?.materials = [anchorMaterialYellow, anchorMaterialPurple, anchorMaterialMagenta]

        self.geometry = geometry
    }

    func setupCone() {
        let geometry: SCNGeometry? = SCNCone(topRadius: boxSize / 2, bottomRadius: boxSize, height: boxSize)
        
        guard let anchorMaterialRed = anchorMaterial(with: .red) else { return }
        guard let anchorMaterialGreen = anchorMaterial(with: .green) else { return }
        guard let anchorMaterialBlue = anchorMaterial(with: .blue) else { return }
        geometry?.materials = [anchorMaterialRed, anchorMaterialGreen, anchorMaterialBlue]

        self.geometry = geometry
    }

    func setupCylinder() {
        let geometry: SCNGeometry? = SCNCylinder(radius: boxSize, height: boxSize)

        guard let anchorMaterialOrange = anchorMaterial(with: .orange) else { return }
        guard let anchorMaterialGray = anchorMaterial(with: .gray) else { return }
        guard let anchorMaterialBlue = anchorMaterial(with: .blue) else { return }
        geometry?.materials = [anchorMaterialOrange, anchorMaterialGray, anchorMaterialBlue]

        self.geometry = geometry
    }

    func setupSphere() {
        let geometry: SCNGeometry? = SCNSphere(radius: boxSize)

        var materialColor: UIColor?

        let col = Int(arc4random()) % 7
        switch col {
            case 0:
                materialColor = UIColor.green
            case 1:
                materialColor = UIColor.red
            case 2:
                materialColor = UIColor.blue
            case 3:
                materialColor = UIColor.green
            case 4:
                materialColor = UIColor.yellow
            case 5:
                materialColor = UIColor.purple
            case 6:
                materialColor = UIColor.magenta
            default:
                break
        }

        guard let anchorMaterial = anchorMaterial(with: materialColor) else { return }
        geometry?.materials = [anchorMaterial]

        self.geometry = geometry
    }

    func setupPyramid() {
        let geometry: SCNGeometry? = SCNPyramid(width: boxSize, height: boxSize, length: boxSize)

        guard let anchorMaterialRed = anchorMaterial(with: .red) else { return }
        guard let anchorMaterialGreen = anchorMaterial(with: .green) else { return }
        guard let anchorMaterialBlue = anchorMaterial(with: .blue) else { return }
        guard let anchorMaterialYellow = anchorMaterial(with: .yellow) else { return }
        guard let anchorMaterialPurple = anchorMaterial(with: .purple) else { return }
        geometry?.materials = [anchorMaterialRed, anchorMaterialGreen, anchorMaterialBlue, anchorMaterialYellow, anchorMaterialPurple]

        self.geometry = geometry
    }

    func setupBox() {
        let geometry: SCNGeometry? = SCNBox(width: boxSize, height: boxSize, length: boxSize, chamferRadius: boxSize / 20)

        guard let anchorMaterialRed = anchorMaterial(with: .red) else { return }
        guard let anchorMaterialGreen = anchorMaterial(with: .green) else { return }
        guard let anchorMaterialBlue = anchorMaterial(with: .blue) else { return }
        guard let anchorMaterialYellow = anchorMaterial(with: .yellow) else { return }
        guard let anchorMaterialPurple = anchorMaterial(with: .purple) else { return }
        guard let anchorMaterialMagenta = anchorMaterial(with: .magenta) else { return }
        geometry?.materials = [anchorMaterialRed, anchorMaterialGreen, anchorMaterialBlue, anchorMaterialYellow, anchorMaterialPurple, anchorMaterialMagenta]

        self.geometry = geometry
    }

    func anchorMaterial(with color: UIColor?) -> SCNMaterial? {
        let material = SCNMaterial()
        material.fillMode = .fill
        material.diffuse.contents = color
        material.specular.contents = UIColor.white
        material.locksAmbientWithDiffuse = true

        return material
    }
}
